### PR TITLE
Use GraphicsMagick

### PR DESCRIPTION
**Pull request by abvarun226**
Reverted the earlier PR Imgur/mandible#111 due to a bug in GM.

When we have a patch for the GM bug, we can merge this
http://graphicsmagick-help.narkive.com/y5AAvruT/gm-help-corrupt-image-when-gm-deals-with-gif
```
$ gm identify error.gifÂ 
gm identify: Corrupt image (error.gif).
gm identify: Request did not return an image.
What is the origin of this GIF file? GM reports that there is a LZW
string table overflow.

There have been exploits where GIF files were updated by virus
software to produce files with oversized LZW tables in order to
compromise the software (e.g. execute code provided in the image file
or crash the program) which opened them.

ImageMagick was re-coded to skip over the extra data in oversized LZW
tables.

In some cases, oversized LZW tables could be due to a bug in the
writing software, or it could be an effort to use larger tables
than allowed by the specification, but in other cases it is
intentional and could be suspect.

```

### CODE REVIEW

**Comment by abvarun226** [comment]
I've patched GM to fix the gif errors, but I need to test it out on a canary instance before merging the PR.

**Comment by briankassouf** [comment]
Do you have the error message from a gif that graphicsmagick failed on? Having to patch graphicsmagick makes it really difficult to run in a non-docker environment. It would be nice to try to avoid this if possible

**Comment by abvarun226** [comment]
The error is
```
00:13:07 0:01 0.000u 32284 gif.c/DecodeImage/290/Coder:
  Excessive LZW string data (string table overflow)
00:13:07 0:01 0.000u 32284 gif.c/DecodeImage/389/CorruptImage:
  Corrupt image (/home/bharghava/Downloads/KW03RdJ.gif)
```

Full debug log,
```
➜  GM /usr/bin/gm identify -debug all ~/Downloads/KW03RdJ.gif 
00:13:07 0:01 0.000u 32284 log.c/SetLogEventMask/1166/Configure:
  Set log event mask: all
00:13:07 0:01 0.000u 32284 blob.c/OpenBlob/2491/Blob:
  Opening blob stream: image 0xb8a1e0, blob 0xb8bd80, mode ReadBinary ...
00:13:07 0:01 0.000u 32284 blob.c/DetachBlob/1118/Blob:
  Detach (reset) blob: blob 0xb8bd80, ref 1
00:13:07 0:01 0.000u 32284 blob.c/OpenBlob/2666/Blob:
    I/O buffer set to 16384 bytes
00:13:07 0:01 0.000u 32284 blob.c/OpenBlob/2688/Blob:
    opened file "/home/bharghava/Downloads/KW03RdJ.gif" as FileStream image 0xb8a1e0, blob 0xb8bd80
00:13:07 0:01 0.000u 32284 blob.c/OpenBlob/2713/Blob:
    read 2053 magic header bytes
00:13:07 0:01 0.000u 32284 blob.c/CloseBlob/869/Blob:
  Closing FileStream blob: image 0xb8a1e0, blob 0xb8bd80, ref 1
00:13:07 0:01 0.000u 32284 blob.c/DetachBlob/1118/Blob:
  Detach (reset) blob: blob 0xb8bd80, ref 1
00:13:07 0:01 0.000u 32284 pixel_cache.c/DestroyCacheInfo/1883/Cache:
  destroy cache 
00:13:07 0:01 0.000u 32284 blob.c/DestroyBlob/1006/Blob:
  Destroy blob (ref counted): image 0xb8a1e0, blob 0xb8bd80, ref 1, filename "/home/bharghava/Downloads/KW03RdJ.gif"
00:13:07 0:01 0.000u 32284 blob.c/DestroyBlob/1021/Blob:
    Destroy blob (real): image 0xb8a1e0, blob 0xb8bd80, ref 0, filename "/home/bharghava/Downloads/KW03RdJ.gif"
00:13:07 0:01 0.000u 32284 constitute.c/ReadImage/1485/Blob:
  Magick=GIF, Filename=/home/bharghava/Downloads/KW03RdJ.gif
00:13:07 0:01 0.000u 32284 constitute.c/ReadImage/1601/Coder:
  Invoking "GIF" decoder (CompuServe graphics interchange format) subimage=0 subrange=0
00:13:07 0:01 0.000u 32284 blob.c/OpenBlob/2491/Blob:
  Opening blob stream: image 0xb8a1e0, blob 0xb8bd80, mode ReadBinary ...
00:13:07 0:01 0.000u 32284 blob.c/DetachBlob/1118/Blob:
  Detach (reset) blob: blob 0xb8bd80, ref 1
00:13:07 0:01 0.000u 32284 blob.c/OpenBlob/2666/Blob:
    I/O buffer set to 16384 bytes
00:13:07 0:01 0.000u 32284 blob.c/OpenBlob/2688/Blob:
    opened file "/home/bharghava/Downloads/KW03RdJ.gif" as FileStream image 0xb8a1e0, blob 0xb8bd80
00:13:07 0:01 0.000u 32284 blob.c/OpenBlob/2713/Blob:
    read 2053 magic header bytes
00:13:07 0:01 0.000u 32284 gif.c/ReadGIFImage/964/Coder:
  Loop extension with iterations 0
00:13:07 0:01 0.000u 32284 resource.c/AcquireMagickResource/230/Resource:
  width +360P/----/256.0MiP
00:13:07 0:01 0.000u 32284 resource.c/AcquireMagickResource/230/Resource:
  height +270P/----/256.0MiP
00:13:07 0:01 0.000u 32284 resource.c/AcquireMagickResource/230/Resource:
  pixels +94.9KiP/----/Unlimited
00:13:07 0:01 0.000u 32284 gif.c/DecodeImage/290/Coder:
  Excessive LZW string data (string table overflow)
00:13:07 0:01 0.000u 32284 gif.c/DecodeImage/389/CorruptImage:
  Corrupt image (/home/bharghava/Downloads/KW03RdJ.gif)
00:13:07 0:01 0.000u 32284 blob.c/DestroyBlob/1006/Blob:
  Destroy blob (ref counted): image 0xb8ec70, blob 0xb91910, ref 1, filename "/home/bharghava/Downloads/KW03RdJ.gif"
00:13:07 0:01 0.000u 32284 blob.c/DestroyBlob/1021/Blob:
    Destroy blob (real): image 0xb8ec70, blob 0xb91910, ref 0, filename "/home/bharghava/Downloads/KW03RdJ.gif"
00:13:07 0:01 0.000u 32284 blob.c/ReferenceBlob/4172/Blob:
  Reference blob: blob 0xb8bd80, ref 2
00:13:07 0:01 0.000u 32284 resource.c/AcquireMagickResource/230/Resource:
  width +29.4KiP/----/256.0MiP
00:13:07 0:01 0.000u 32284 resource.c/AcquireMagickResource/230/Resource:
  height +5.6KiP/----/256.0MiP
00:13:07 0:01 0.000u 32284 resource.c/AcquireMagickResource/230/Resource:
  pixels +165.3MiP/----/Unlimited
00:13:07 0:01 0.000u 32284 gif.c/DecodeImage/145/CorruptImage:
  Corrupt image (/home/bharghava/Downloads/KW03RdJ.gif)
00:13:07 0:01 0.000u 32284 blob.c/CloseBlob/869/Blob:
  Closing FileStream blob: image 0xb8a1e0, blob 0xb8bd80, ref 2
00:13:07 0:01 0.000u 32284 blob.c/DetachBlob/1118/Blob:
  Detach (reset) blob: blob 0xb8bd80, ref 2
00:13:07 0:01 0.000u 32284 constitute.c/ReadImage/1612/Coder:
  Returned from "GIF" decoder: frames=2 cache=present monochrome=False grayscale=False class=PseudoClass colorspace=RGB
00:13:07 0:01 0.000u 32284 pixel_cache.c/DestroyCacheInfo/1883/Cache:
  destroy cache /home/bharghava/Downloads/KW03RdJ.gif[0]
00:13:07 0:01 0.000u 32284 blob.c/DestroyBlob/1006/Blob:
  Destroy blob (ref counted): image 0xb8a1e0, blob 0xb8bd80, ref 2, filename "/home/bharghava/Downloads/KW03RdJ.gif"
00:13:07 0:01 0.000u 32284 pixel_cache.c/DestroyCacheInfo/1883/Cache:
  destroy cache 
00:13:07 0:01 0.000u 32284 blob.c/DestroyBlob/1006/Blob:
  Destroy blob (ref counted): image 0xb8ec70, blob 0xb8bd80, ref 1, filename "/home/bharghava/Downloads/KW03RdJ.gif"
00:13:07 0:01 0.000u 32284 blob.c/DestroyBlob/1021/Blob:
    Destroy blob (real): image 0xb8ec70, blob 0xb8bd80, ref 0, filename "/home/bharghava/Downloads/KW03RdJ.gif"
/usr/bin/gm identify: Corrupt image (/home/bharghava/Downloads/KW03RdJ.gif).
00:13:07 0:01 0.000u 32284 command.c/IdentifyImageCommand/8636/Option:
  Request did not return an image
/usr/bin/gm identify: Request did not return an image.
00:13:07 0:01 0.000u 32284 magick.c/DestroyMagick/170/Configure:
  Destroy Magick
```

**Comment by pfischermx** [comment]
How are we building the gm? The diff says 'gm' is a binary but if we are building it maybe we should have it build inside the Dockerfile so that people can know the options (or add) we used to build it.

**Comment by briankassouf** [comment]
@abvarun226 What version of GraphicsMagick is being used? This tread talks about this error being fixed with a 64-bit version http://graphicsmagick-help.narkive.com/PKlF8JGC/gm-help-gif-read-error

**Comment by abvarun226** [comment]
It is GraphicsMagick 1.3.25

The changelog also says the LZW overflow issue has been addressed, but the issue is still there in 1.3.25
```
2016-11-26 Bob Friesenhahn <bfriesen@simple.dallas.tx.us>

coders/gif.c (DecodeImage): Applied fixes by Tianyu Lang for "Excessive LZW string data" problem leading to "Corrupt image" report while reading some GIF files.
```
http://www.graphicsmagick.org/ChangeLog-2016.html

**Comment by pfischermx** [comment]
How are we building the binary? :)

**Comment by abvarun226** [comment]
The binary is built locally and not inside Docker. I'll look at building in inside Docker tomorrow.

**Comment by abvarun226** [comment]
I just downloaded the source from mercurial and built the binary with source at tip. Seems like the issue is resolved with the latest source. I don't think its been packaged yet and available on debian repo.

**Comment by abvarun226** [comment]
Made some changes to build GM from latest source inside docker itself with all feature support,

```

Feature Support:
  Native Thread Safe       yes
  Large Files (> 32 bit)   yes
  Large Memory (> 32 bit)  yes
  BZIP                     yes
  DPS                      no
  FlashPix                 no
  FreeType                 yes
  Ghostscript (Library)    no
  JBIG                     yes
  JPEG-2000                yes
  JPEG                     yes
  Little CMS               yes
  Loadable Modules         no
  OpenMP                   yes (201511)
  PNG                      yes
  TIFF                     yes
  TRIO                     no
  UMEM                     no
  WebP                     yes
  WMF                      yes
  X11                      yes
  XML                      yes
  ZLIB                     yes

Host type: x86_64-pc-linux-gnu

Configured using the command:
  ./configure  '--build' 'x86_64-linux-gnu' '--enable-shared' '--enable-static' '--enable-libtool-verbose' '--prefix=/usr' '--mandir=${prefix}/share/man' '--infodir=${prefix}/share/info' '--docdir=${prefix}/share/doc/graphicsmagick' '--with-gs-font-dir=/usr/share/fonts/type1/gsfonts' '--with-x' '--x-includes=/usr/include/X11' '--x-libraries=/usr/lib/X11' '--without-dps' '--without-modules' '--without-frozenpaths' '--with-webp' '--with-perl' '--with-perl-options=INSTALLDIRS=vendor' '--enable-quantum-library-names' '--with-quantum-depth=16' 'build_alias=x86_64-linux-gnu' 'CFLAGS=-Wall -g -fno-strict-aliasing -O2' 'LDFLAGS=' 'CXXFLAGS=-Wall -g -fno-strict-aliasing -O2' 'CC=gcc' 'LIBS=-ljbig -lwebp -llcms2 -ltiff -lfreetype -ljpeg -lpng16 -lwmflite -lXext -lSM -lICE -lX11 -llzma -lbz2 -lxml2 -lz -lm -lgomp -lpthread' 'CPPFLAGS=-I/usr/include/X11 -I/usr/include/freetype2 -I/usr/include/libxml2' 'CXX=g++'

Final Build Parameters:
  CC       = gcc
  CFLAGS   = -fopenmp -Wall -g -fno-strict-aliasing -O2 -Wall
  CPPFLAGS = -I/usr/include/X11 -I/usr/include/freetype2 -I/usr/include/libxml2 -I/usr/include/X11 -I/usr/include/freetype2
  CXX      = g++
  CXXFLAGS = -Wall -g -fno-strict-aliasing -O2
  LDFLAGS  = -L/usr/lib/X11 -L/usr/lib/x86_64-linux-gnu
  LIBS     = -ljbig -lwebp -llcms2 -ltiff -lfreetype -ljpeg -lpng16 -lwmflite -lXext -lSM -lICE -lX11 -llzma -lbz2 -lxml2 -lz -lm -lgomp -lpthread -ljbig -lwebp -llcms2 -ltiff -lfreetype -ljasper -ljpeg -lpng16 -lwmflite -lXext -lSM -lICE -lX11 -llzma -lbz2 -lxml2 -lz -lm -lgomp -lpthread
```

**Comment by abvarun226** [comment]
Tested on canary instance and it looks good.

**Comment by abvarun226** [comment]
@pastudan @briankassouf @pfischermx 
Are we good to merge this?

**Review by pastudan — APPROVED** [review]

**Comment by abvarun226** [comment]
@pastudan @briankassouf 
Can we revisit this? The new version of GM with the fix is still not out. We will still need to compile from the source.

File: imageprocessor/ocr_test.go [hunk 1]
```diff
@@ -25,20 +25,20 @@ func TestStandardOCR(t *testing.T) {
...
+// 	}
+// 	defer image.Clean()
+//
+// 	ocrStratagy := DuelOCRStratagy()
+// 	ocrStratagy.Process(image)
+//
+// 	if image.GetOCRText() != "hello" {
+// 		t.Fatalf("Did not get proper Duel OCR text back %s != hello", image.GetOCRText())
+// 	}
+// }
```
Comment (pastudan) [inline imageprocessor/ocr_test.go]: Lets go ahead and remove. We can restore from git if we decide to bring this feature back

**Review by pastudan — CHANGES_REQUESTED** [review]

File: imageprocessor/ocr_test.go (hunk 1 above)
Comment (abvarun226) [inline imageprocessor/ocr_test.go]: Yep, true. Updated the PR

**Review by pastudan — APPROVED** [review]

**Review by kodermike — APPROVED** [review]

**Pull request merged**